EIPS/eip-00.md

```diff
@@ -0,0 +1,790 @@
+
+The old `nonce` account parameter remains in use for transactions initiated by EOAs and for the `CREATE` opcode.
+
+#### NonceManager Pseudocode
+
+```
+
+if evm.caller == AA_ENTRY_POINT:
+    validate_increment()
+else:
+    get()
+
+def get():
+    if len(evm.calldata) != 44:
+        evm.revert()
+
+    // address sender, uint192 key
+    address = to_uint160_be(evm.calldata[0:20])
+    key = to_uint192_be(evm.calldata[20:44])
+
+    nonce = storage.get(keccack(address, key))
+
+    evm.return((key << 64) + nonce)
+
+def validate_increment():
+
+    address = to_uint160_be(evm.calldata[0:20])
+    key = to_uint192_be(evm.calldata[20:44])
+    nonce = to_uint64_be(evm.calldata[44:52])
+
+    current_nonce = storage.get(keccack(address, key))
+
+    if (nonce != current_nonce):
+        evm.revert()
+
+    storage.set(keccack(address, key), current_nonce + 1)
+
+```
+
+#### NonceManager Bytecode and deployment
+
+TODO.
+
+### Gas fees are charged directly from the contract balance
+
+The maximum gas cost of the AA_TX_TYPE transaction is defined as:
+
+```
+
+maxPossibleGasCost = AA_BASE_GAS_COST +
+  callGasLimit +
+  paymasterGasLimit +
+  validationGasLimit
+
+```
+
+If `paymaster` is not specified, the `maxPossibleGasCost` is charged up-front, before any computation is done in any
+execution frame, from the balance of the `sender` address.
+If `paymaster` is specified, the gas cost is charged from its balance.
+The transaction is invalid if the balance of the account that is being pre-charged,
+whether it is a `sender` or a `paymaster`, is insufficient.
+After the transaction finishes its execution, the address that was pre-charged may receive a gas refund.
+
+### Gas fees charged for transaction input
+
+For all the existing transaction types, G_txdatazero (4 gas) and G_txdatanonzero (16 gas) per byte is
+charged for the `data` parameter.
+
+Transaction Type AA_TX_TYPE introduces the following dynamic length inputs: `callData`, `paymasterData`,
+`deployerData`, `signature`. Each of these parameters' gas cost is counted towards transaction data cost.
+This transaction data gas cost is referred to as `calldataCost` and is subtracted from the `callGasLimit`
+before execution of the transaction.
+The transaction is considered INVALID if `callGasLimit` is smaller than `calldataCost`.
+
+### Builder Fee
+
+As we need to account for an additional off-chain work that block builders have to perform to
+include `AA_TX_TYPE` transactions in their blocks, as well as a potential L1 gas cost for builders
+operating on L2 rollups, and given that this work does not correspond to the amount of gas spent on
+validation and is not linked to the gas price, the `sender` may decide
+to pay an extra `builderFee` as a "tip" to the block builder.
+
+This value is denominated in wei and is passed from the `sender` to the `coinbase` as part of the gas pre-charge.
```

Comment:
Thought about this more, I understand why this concept is needed for full gas abstraction on L2s.

However, I'd like to know what costs can contribute to this field that are not better charged by the VM itself can modeled as a limit. Also, `callGasLimit` would be a better place to subtract these costs from, so that `validationGasLimit` can stay relatively static for a particular account's validation path.

On Arbitrum, for example, my understanding the VM charges the L1 calldata costs by subtracting from the initial `gasLimit`. That same concept could be applied here, and across L2s that already have ways of charging for this built into their VMs.

Its pretty important to model this as a limit, especially when it comes to calldata. We've found that users who want to ensure that their `preVerificationGas` is always valid while L2 fees are changing quickly, need to increase their `preVerificationGas` by up to 25%. Since L1 calldata costs are the majority of the cost of a L2 transaction, this basically makes using 4337 20%+ more expensive.

Reply:
`builderFee` is a generalization of `preVerificationGas`.  It's deliberately denominated in wei rather than gas, because it accounts for costs not directly associated with the amount of gas consumed by the transaction.

One example is aggregation fees, especially when using a Snark aggregator with heavy proving costs - a cost that differs between types of aggregators.  Another is L1 calldata gas for rollups, which may or may not be possible to calculate via the L2's gas oracle, depending on each L2 architecture.  Yet another one is ZK rollup proving costs involving the number of keccaks in the transaction.

Some L2 networks are actually considering a new transaction standard for multidimensional gas to address their individual cost dependencies.

On chains where everything can be calculated as part of the transaction gas, such as on Ethereum mainnet, this value may always be zero.  But on other chains there will always be costs not associated with gas, that the user may not be able to calculate.  The sequencer (or RPC node) will have to propose a sensible value based on the information it has (which the user might not have, such as the GPU cost of proving the transaction, combined with the number of transactions in the next batch to be proven).

I hope this value can be zero in as many cases as possible, but I think we should have it in the standard to support the cases where it's needed.

Do you see a way to model this as a limit, while still supporting every such case of off-chain costs associated with inclusion of the transaction?  The above are just a few examples.  We need this value unless we can come up with a limit-based model that would work with every possible future off-chain cost.

Reply:
I think its very important to model costs that may be time-variable in a limit field. Else users are stuck paying extra to achieve reliable time to mine.

The solution here may vary from network to network. On Arbitrum or Optimism the protocol could easily be modified to account for L1 calldata costs in the `validationGasLimit` field. Any network that charges variable off-chain fees to their users can just convert those fees to gas by dividing by the base fee, and then subtract that gas from the `validationGasLimit`.

I could see `builderFee` being used to charge a static fee for offchain signature aggregator proving costs. Then `validationGasLimit` can be used to cover the variable aggregated signature onchain validation cost (which will reduce with the number of aggregated signatures).

I'm not against the `builderFee` field existing, but I hope as this EIP is adopted onto chains they utilize limit fields to charge for variable costs.

Reply:
@drortirosh I'm not sure I understand how it would work, when it comes to offchain costs such as ZK proving costs.

For certain use cases it could work, if they are a part of that chain's consensus.  E.g. the chain's proving cost could take into account things like the number of keccaks in a transaction or the number of transactions in a batch.  This will be a part of the proof since it's a part of that chain's consensus.

But for use cases involving a 3rd party, how would it work?  For example, when we add the aggregated native AA EIP, the signature prover may be a 3rd party service that uses an aggregation method the network is unaware of.  The consensus won't be aware of the off-chain costs associated with every possible aggregation algorithm.  Do you see a way to model this as a limit?

However, even if the answer is no, I agree that it would be best to model as many things as possible as a limit, and only resort to a static fee when there's absolutely no other way - which should be quite rare.

@dancoombs @drortirosh do you think we should add two dimensions here, enabling both a `builderFeeLimit` and a `builderStaticFee`, or do you see a way we could completely avoid static fees for every possible case?

Reply:
I think it is very important to allow users to express a limit value for values that can vary (quickly relative to block time) and are part of consensus. The two main use cases being L2->L1 calldata costs and dynamic portions of aggregation validation. I think at a minimum there should be a `builderFeeLimit` field denominated in wei.

If we are convinced that there are values outside of consensus that are needed (your explanation makes sense, but I haven't seen any of these use cases on a live implementation), I'd prefer the two dimensions.

We should make it very clear to implementors that using `builderStaticFee` for time-varying fees is bad practice and will force the users to incur extra costs in order to ensure their transaction gets mined.